src/mantle/testing/concerns/snapshots/trait-mocks-http-requests.php

```diff
@@ -0,0 +1,194 @@
+	 */
+	public bool $snapshot = false;
+
+	/**
+	 * Path to snapshot file.
+	 */
+	private string $snapshot_file;
+
+	/**
+	 * Internal Request instance.
+	 */
+	private Request $request;
+
+	/**
+	 * Generate a response from a snapshot file.
+	 */
+	public function with_snapshot(): static {
+		$this->snapshot = true;
+
+		return $this;
+	}
+
+	/**
+	 * Fetch the snapshot from storage or make an actual request.
+	 *
+	 * @throws InvalidArgumentException Thrown when called without snapshot being set to true.
+	 *
+	 * @param Request $request Current request object.
+	 */
+	public function process_from_snapshot( Request $request ): ?array {
+		if ( ! $this->snapshot ) {
+			throw new InvalidArgumentException( 'Snapshot not enabled for mocked request.' );
+		}
+
+		$this->request       = $request;
+		$this->snapshot_file = $this->get_snapshot_path( $request );
+
+		if ( $this->should_update_snapshots() || ! file_exists( $this->snapshot_file ) ) {
+			if ( Utils::is_ci() ) {
+				$this->get_test_case()->fail(
+					'Snapshot does not exist for a request that is being mocked with a snapshot: ' . $request->url(),
+				);
+			}
+
+			// Add the filter that will capture the HTTP response for the snapshot.
+			add_filter( 'http_response', [ $this, 'capture_http_response_for_snapshot' ], 10, 3 );
+
+			return null;
+		}
+
+		$contents = wp_json_file_decode( $this->snapshot_file, [ 'associative' => true ] );
+
+		if ( ! is_array( $contents ) ) {
+			Utils::error( 'Snapshot file is not valid JSON: ' . $this->snapshot_file, 'HTTP Requests' );
+
+			return null;
+		}
+
+		return $contents;
+	}
+
+	/**
+	 * Get the test case instance.
+	 */
+	private function get_test_case(): TestCase {
+		return Container::get_instance()->make( TestCase::class );
+	}
+
+	/**
+	 * Retrieve the snapshot path for the request.
+	 *
+	 * @param Request $request Request object.
+	 */
+	private function get_snapshot_path( Request $request ): string {
+		return $this->get_snapshot_directory() . DIRECTORY_SEPARATOR . $this->get_snapshot_id( $request ) . '.json';
+	}
+
+	/**
+	 * Determines the directory where snapshots are stored.
+	 *
+	 * By default a `__http_snapshots__` directory is created at the same level as
+	 * the test class.
+	 */
+	private function get_snapshot_directory(): string {
+		$reflection = new ReflectionClass( $this->get_test_case() );
+
+		return collect( [
+			dirname( $reflection->getFileName() ),
+			'__http_snapshots__',
+			$reflection->getShortName(),
+		] )->join( DIRECTORY_SEPARATOR );
+	}
+
+	/**
+	 * Determines the snapshot's id. By default, the test case's class and
+	 * method names are used.
+	 *
+	 * @param Request $request
+	 */
+	private function get_snapshot_id( Request $request ): string {
+		$test_case = $this->get_test_case();
+
+		return collect( [
+			$test_case->nameWithDataSet(),
+			$request->enum_method()->value,
+			str_replace( [ '/', ':', DIRECTORY_SEPARATOR ], '-', $request->url() ),
```

Comment:
This will work for oEmbed requests, but wouldn't work for requests that go to a generic endpoint that accepts arguments (e.g., a REST API or GraphQL endpoint that accepts POST arguments or a JSON payload). Would it be possible to make a hash out of the request (e.g., URL + HTTP verb + headers + body)? This may be trivial to do since under the hood WP is building a request object that has this information in it already - it could be converted to a string and hashed, or the relevant pieces could be JSON-encoded and hashed, or similar. If we want to preserve the ability to identify what tests the snapshot is associated with, we could keep the test case class and method names, and append the hash to the end rather than a representation of the URL.